docs/root/start/sandboxes/single-page-app.rst

```diff
@@ -0,0 +1,597 @@
+    single-page-app-envoy-1       single-page-app-envoy       "/docker-entrypoint.sh envoy -c /etc/envoy/envoy.yaml ..." envoy     2 minutes ago   Up 2 minutes           0.0.0.0:10000-10001->10000-10001/tcp, :::10000-10001->10000-10001/tcp
+    single-page-app-myhub-1       single-page-app-myhub       "/opt/myhub/app.py"                                        myhub     2 minutes ago   Up 2 minutes (healthy) 0.0.0.0:7000->7000/tcp, :::7000->7000/tcp
+    single-page-app-myhub-api-1   single-page-app-myhub-api   "/opt/myhub/app.py"                                        myhub-api 2 minutes ago   Up 2 minutes (healthy)
+    single-page-app-ui-1          single-page-app-ui          "/entrypoint.sh dev.sh"                                    ui        2 minutes ago   Up 2 minutes (healthy)
+
+.. _install_sandboxes_single_page_app_step_login:
+
+Step 3: Browse to the dev app and login
+***************************************
+
+The development app should now be available at http://localhost:10001 and provide a login button:
+
+.. image:: /start/sandboxes/_static/spa-login.png
+   :align: center
+
+.. note::
+   The dummy OAuth provider automatically trusts everyone as a hard-coded ``envoydemo`` user and redirects back to the app.
+
+   In a real world scenario the provider would authenticate and authorize the user before proceeding.
+
+The sandbox is configured with an inverted match on
+:ref:`pass_through_matcher <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.pass_through_matcher>`.
+
+This ignores all paths for OAuth other than:
+
+- ``/authorize.*``
+- ``/hub.*``
+- ``/login``
+- ``/logout``.
+
+.. literalinclude:: _include/single-page-app/envoy.yml
+   :language: yaml
+   :lines: 37-46
+   :linenos:
+   :lineno-start: 37
+   :emphasize-lines: 3-8
+   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`
+
+When a user clicks ``login`` the app initiates the OAuth flow by calling the ``/login`` path in Envoy.
+
+This redirects the user to the OAuth provider for authorization/authentication with a further redirect link:
+
+.. literalinclude:: _include/single-page-app/envoy.yml
+   :language: yaml
+   :lines: 34-39
+   :linenos:
+   :lineno-start: 34
+   :emphasize-lines: 3-4
+   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`
+
+On successful authorization/authentication the user is redirected back via this link to the app with the necessary OAuth code to proceed:
+
+.. literalinclude:: _include/single-page-app/envoy.yml
+   :language: yaml
+   :lines: 44-50
+   :linenos:
+   :lineno-start: 44
+   :emphasize-lines: 3-5
+   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`
+
+Envoy then uses this code with its client secret to generate an access token for the user, which Envoy then retains:
```

Comment:
I'm not sure I understand this portion actually. The token is being retained in memory right? Is this on the response path? Kind of confused.

Reply:
this is how envoy/oauth works - after authorization user returns and makes a requiest including `code`

oauth server issues token to envoy after client makes request including `code`

wondering how to make this cleaerer - open to suggestions

Reply:
following up offline conv - the formal name is "authorization code" https://oauth.net/2/grant-types/authorization-code/

i will update terminology in pr

Reply:
done